src/gf.c

```diff
@@ -2526,7 +2534,45 @@ jl_code_instance_t *jl_compile_method_internal(jl_method_instance_t *mi, size_t
         }
     }
 
-    codeinst = jl_generate_fptr(mi, world);
+    // Ok, compilation is enabled. We'll need to try to compile something (probably).
+    // Try to find a codeinst we have already inferred (e.g. while we were compiling
+    // something else).
+    codeinst = jl_method_inferred_with_source(mi, world);
+
+    // Everything from here on our is considered (user facing) compile time
+    uint64_t start = jl_typeinf_timing_begin();
+    int is_recompile = jl_atomic_load_relaxed(&mi->cache) != NULL;
+
+    // This codeinst hasn't been previously inferred do that now
+    if (!codeinst) {
+        // Don't bother inferring toplevel thunks or macros - the performance cost of inference is likely
+        // to significantly exceed the actual runtime.
+        int should_skip_inference = !jl_is_method(mi->def.method) || jl_symbol_name(mi->def.method->name)[0] == '@';
+
+        if (!should_skip_inference) {
+            codeinst = jl_type_infer(mi, world, 0, SOURCE_MODE_ABI);
+        }
+    }
+
+    if (codeinst) {
+        // Something went very wrong. Bail out.
+        if (jl_atomic_load_acquire(&codeinst->invoke) != NULL) {
```

Comment:
```suggestion
        if (jl_code_instance_t *ci = jl_method_compiled(mi, world))
            return ci; // compiled while we were away
```

(I am only 10% confident this C++-ism is valid in C though, so YMMV)

Reply:
I don't think it makes sense to re-walk the CodeInstance list here. The only time when it would make sense is after acquiring the codegen lock, but we don't have it yet. This check is mostly just for whether this is something that was set by inference (likely constabi).